src/libraries/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -416,6 +426,265 @@ public static unsafe void Copy(Array sourceArray, int sourceIndex, Array destina
             // Less common
             CopyImpl(sourceArray!, sourceIndex, destinationArray!, destinationIndex, length, reliable: false);
         }
+
+        private static unsafe void CopyImpl(Array sourceArray, int sourceIndex, Array destinationArray, int destinationIndex, int length, bool reliable)
+        {
+            if (sourceArray == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.sourceArray);
+            if (destinationArray == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.destinationArray);
+
+            if (sourceArray.GetType() != destinationArray.GetType() && sourceArray.Rank != destinationArray.Rank)
+                throw new RankException(SR.Rank_MustMatch);
+
+            ArgumentOutOfRangeException.ThrowIfNegative(length);
+
+            int srcLB = sourceArray.GetLowerBound(0);
+            ArgumentOutOfRangeException.ThrowIfLessThan(sourceIndex, srcLB);
+            ArgumentOutOfRangeException.ThrowIfNegative(sourceIndex - srcLB, nameof(sourceIndex));
+            sourceIndex -= srcLB;
+
+            int dstLB = destinationArray.GetLowerBound(0);
+            ArgumentOutOfRangeException.ThrowIfLessThan(destinationIndex, dstLB);
+            ArgumentOutOfRangeException.ThrowIfNegative(destinationIndex - dstLB, nameof(destinationIndex));
+            destinationIndex -= dstLB;
+
+            if ((uint)(sourceIndex + length) > sourceArray.NativeLength)
+                throw new ArgumentException(SR.Arg_LongerThanSrcArray, nameof(sourceArray));
+            if ((uint)(destinationIndex + length) > destinationArray.NativeLength)
+                throw new ArgumentException(SR.Arg_LongerThanDestArray, nameof(destinationArray));
+
+            ArrayAssignType assignType = ArrayAssignType.WrongType;
```

Comment:
```suggestion
            ArrayAssignType assignType;
```
Should not be needed.